### DISCUSSION TITLE
[Core Protocol] Can‘t find the  interacting case of step() function in the fault proof distribute game

### DISCUSSION THREAD

**Discussion by joohhnnn** [ethereum-optimism/developers#498] [Core Protocol] [answered]
### Did you check the documentation?

- [X] Yes

### Did you read the specs?

- [X] Yes
- [ ] No

### Did you check for duplicate questions?

- [X] Yes

### Issue Description

**Issue Background:** 
In the fault proof system, when a dispute reaches the finest granularity, the `step()` function in the **Game contract** will be called. This function indirectly invokes the **MIPS** virtual machine for verification. Before this verification process begins, it is necessary to call the `addLocalData()` function within the **Game contract**. This function adds data required by the **MIPS** virtual machine during the verification process to the PreImageOracle contract.

**Issue:**
I'm looking to understand the` step()` function's usage in real-world production or testing environments, but I **haven't found** any use cases, not even on testnets. Additionally, I've come across several **PreImageOracle contracts** deployed at different times, and none of them seem to have a direct or indirect `loadLocalData()` function call.

Here are some PreImageOracle contracts I looked at on various testnets, and none of them seem to have the relevant function calls:

[Contract created 92 days ago](https://sepolia.etherscan.io/address/0x98b9D322bBce586259283F79aF835744777E88c3)
[Contract created 55 days ago](https://sepolia.etherscan.io/address/0x3f62614ff1194F01dfaF6d027951c43F9878F92d)
[Contract created 36 days ago](https://sepolia.etherscan.io/address/0x627F825CBd48c4102d36f287be71f4234426b9e4)

**Comment by sbvegan** [comment]
Hey @joohhnnn, I flagged this internally and hope to have an answer for you soon.

**Comment by ajsutton** [comment, marked as answer]
It's not necessarily required to call `loadLocalData()` or to load any data into the preimage oracle when calling `step` - it's only required if the single MIPS CPU instruction being executed in the `step` call is actually reading preimage data and then only that one preimage needs to be uploaded.  `op-challenger` will automatically upload the required preimage data as part of its call to step when necessary (https://github.com/ethereum-optimism/optimism/blob/78e1084ec14d3003cb9e546b9eb5a22db7408ac2/op-challenger/game/fault/responder/responder.go#L90-L129 is probably the best place to start for that code).

I believe it costs around 600ETH in total bonds to get a dispute game down to where the `step` function is called so it is an exceptionally unusual thing to have happen. You're right that it hasn't happened on any of the `DisputeGameFactory` instances used by the `OptimismPortal`. There are a number of cases around on public chains but I'm not sure I could find them again now and I'm not sure if they wound up needing to call `addLocalData`.  We do have e2e tests that do this on local devnets to test it - including specifically targeting steps that require specific types of preimage data. We can also use `eth_Call` to invoke `step` without having to push a dispute game all the way down.
